File: test-src/33_complex.kt
fun foo(a: String, b: Map<String, String>, c: ((String) -> Unit)? = null): String? {
  c?.invoke("a=$a")
  return a
}

fun fooFunc(a: String, b: Map<String, String>, c: ((String) -> Unit)? = null): String? {
  return null
}

fun run(): String {
  var outerValue = ""
  var a = foo("bar", b = HashMap<String, String>(), c = { value -> outerValue = value }) ?: return "fail1"
  a = foo("bar", b = HashMap<String, String>(), c = { value -> outerValue = value })
  fooFunc("bar", b = HashMap<String, String>(), c = { value -> outerValue = value })

  if (a != "bar") {
    println("fail: a=$a (should be bar)")
    return "fail2"
  }

  if (outerValue != "a=bar") {
    println("fail: outerValue=$outerValue (should be a=bar)")
    return "fail3"
  }

  return "ok"
}

fun main(args: Array<String>) {
  if (run() != "ok") {
    println("fail: invalid return")
  } else {
    println("success")
  }
}

File: test-dest/33_complex.swift
func foo(a: String, b: [String: String], c: ((String) -> Void)? = nil) -> String? {
  c?("a=\(a)")
  return a
}

func fooFunc(a: String, b: [String: String], c: ((String) -> Void)? = nil) -> String? {
  return nil
}

func run() -> String {
  var outerValue = ""
  let _kotliftOptional1 = foo("bar", b: [String: String](), c: { value in outerValue: value }); if _kotliftOptional1 == nil { return "fail1" }; var a = _kotliftOptional1!
  a = foo("bar", b: [String: String](), c: { value in outerValue: value })
  fooFunc("bar", b: [String: String](), c: { value in outerValue: value })

  if a != "bar" {
    print("fail: a=\(a) (should be bar)")
    return "fail2"
  }

  if outerValue != "a=bar" {
    print("fail: outerValue=\(outerValue) (should be a=bar)")
    return "fail3"
  }

  return "ok"
}

func main(args: [String]) {
  if run() != "ok" {
    print("fail: invalid return")
  } else {
    print("success")
  }
}

main([])
